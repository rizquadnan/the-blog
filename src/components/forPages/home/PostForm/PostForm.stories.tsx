import * as React from 'react'
import { Meta, Story } from '@storybook/react'
import { PostForm } from './PostForm'

type TPostForm = React.ComponentProps<typeof PostForm>
export default {
  title: 'User Inputs/PostForm',
  component: PostForm,
} as Meta<TPostForm>

const Template: Story<TPostForm> = (args) => <PostForm {...args} />

export const Default = Template.bind({})
Default.args = {
  author: String(271030),
}
