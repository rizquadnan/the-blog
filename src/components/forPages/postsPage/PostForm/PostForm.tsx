import { Button } from '@chakra-ui/button'
import {
  FormControl,
  FormErrorMessage,
  FormLabel,
} from '@chakra-ui/form-control'
import React from 'react'
import { useForm } from 'react-hook-form'
import { VStack } from '@/components/uiKit/VStack'
import { Text, Textarea } from '@chakra-ui/react'

type TFormValues = {
  title: string
  body: string
}
type TPostFormCreate = {
  variant: 'create'
  author: string
  onSubmit: (props: TFormValues) => void
}
type TPostFormUpdate = {
  variant: 'update'
  initialValues: TFormValues & { post: string }
  onSubmit: (props: TFormValues) => void
}
type TPostForm = (TPostFormCreate | TPostFormUpdate) & {
  author: string
  isSubmitLoading?: boolean
}

export function PostForm(props: TPostForm) {
  const {
    register,
    handleSubmit,
    formState: { errors, isDirty },
  } = useForm<TFormValues>({
    defaultValues:
      props.variant === 'update'
        ? props.initialValues
        : { body: '', title: '' },
  })

  const notEligbleToUpdate = props.variant === 'update' && !isDirty
  const formTitle =
    props.variant === 'create'
      ? `Creating Post for user: ${props.author}`
      : `Updating Post: ${props.initialValues.post}`

  return (
    <VStack as="form" minW="280px">
      <Text fontSize="18px" fontWeight={600}>
        {formTitle}
      </Text>
      <FormControl isInvalid={Boolean(errors.title)}>
        <FormLabel>Title</FormLabel>
        <Textarea
          {...register('title', {
            required: 'Title is required',
            minLength: {
              value: 3,
              message: 'Title must be at least 3 characters',
            },
          })}
          placeholder="Title"
        />
        <FormErrorMessage>{errors.title?.message}</FormErrorMessage>
      </FormControl>

      <FormControl isInvalid={Boolean(errors.body)}>
        <FormLabel>Body</FormLabel>
        <Textarea
          {...register('body', {
            required: 'Body is required',
            minLength: {
              value: 10,
              message: 'Body must be at least 10 characters',
            },
          })}
          placeholder="Body"
        />
        <FormErrorMessage>{errors.body?.message}</FormErrorMessage>
      </FormControl>
      <Button
        w="100%"
        mt="48px !important"
        isDisabled={notEligbleToUpdate}
        isLoading={props.isSubmitLoading}
        onClick={handleSubmit(props.onSubmit)}
      >
        Submit
      </Button>
    </VStack>
  )
}

export default PostForm
