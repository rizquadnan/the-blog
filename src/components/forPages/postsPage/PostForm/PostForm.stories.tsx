import * as React from 'react'
import { Meta, Story } from '@storybook/react'
import { PostForm } from './PostForm'

type TPostForm = React.ComponentProps<typeof PostForm>
export default {
  title: 'User Inputs/PostForm',
  component: PostForm,
} as Meta<TPostForm>

const Template: Story<TPostForm> = (args) => <PostForm {...args} />

export const CreateVariant = Template.bind({})
CreateVariant.args = {
  author: String(271030),
  variant: 'create',
}

export const UpdateVariant = Template.bind({})
UpdateVariant.args = {
  variant: 'update',
  initialValues: {
    body: 'Some Body',
    title: 'Some Title',
    post: String(241232),
  },
}
