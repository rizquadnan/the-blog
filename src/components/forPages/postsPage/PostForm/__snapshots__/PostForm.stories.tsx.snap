// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`User Inputs/PostForm CreateVariant smoke-test 1`] = `
Object {
  "checked": undefined,
  "children": Array [
    Object {
      "checked": undefined,
      "children": undefined,
      "name": "Creating Post for user: 271030",
      "pressed": undefined,
      "role": "text",
      "value": undefined,
    },
    Object {
      "checked": undefined,
      "children": undefined,
      "name": "Title",
      "pressed": undefined,
      "role": "text",
      "value": undefined,
    },
    Object {
      "checked": undefined,
      "children": undefined,
      "multiline": true,
      "name": "Title",
      "pressed": undefined,
      "role": "textbox",
      "value": undefined,
    },
    Object {
      "checked": undefined,
      "children": undefined,
      "name": "Body",
      "pressed": undefined,
      "role": "text",
      "value": undefined,
    },
    Object {
      "checked": undefined,
      "children": undefined,
      "multiline": true,
      "name": "Body",
      "pressed": undefined,
      "role": "textbox",
      "value": undefined,
    },
    Object {
      "checked": undefined,
      "children": undefined,
      "name": "Submit",
      "pressed": undefined,
      "role": "button",
      "value": undefined,
    },
  ],
  "name": "user-inputs-postform--create-variant",
  "pressed": undefined,
  "role": "WebArea",
  "value": undefined,
}
`;

exports[`User Inputs/PostForm UpdateVariant smoke-test 1`] = `
Object {
  "checked": undefined,
  "children": Array [
    Object {
      "checked": undefined,
      "children": undefined,
      "name": "Updating Post: 241232",
      "pressed": undefined,
      "role": "text",
      "value": undefined,
    },
    Object {
      "checked": undefined,
      "children": undefined,
      "name": "Title",
      "pressed": undefined,
      "role": "text",
      "value": undefined,
    },
    Object {
      "checked": undefined,
      "children": undefined,
      "multiline": true,
      "name": "Title",
      "pressed": undefined,
      "role": "textbox",
      "value": "Some Title",
    },
    Object {
      "checked": undefined,
      "children": undefined,
      "name": "Body",
      "pressed": undefined,
      "role": "text",
      "value": undefined,
    },
    Object {
      "checked": undefined,
      "children": undefined,
      "multiline": true,
      "name": "Body",
      "pressed": undefined,
      "role": "textbox",
      "value": "Some Body",
    },
    Object {
      "checked": undefined,
      "children": undefined,
      "disabled": true,
      "name": "Submit",
      "pressed": undefined,
      "role": "button",
      "value": undefined,
    },
  ],
  "name": "user-inputs-postform--update-variant",
  "pressed": undefined,
  "role": "WebArea",
  "value": undefined,
}
`;
